// @pixi-spine/loader-uni/lib/loader-uni.js
/* eslint-disable */
 
/*!
 * @pixi-spine/loader-uni - v3.0.5
 * Compiled Wed, 14 Jul 2021 08:35:24 UTC
 *
 * @pixi-spine/loader-uni is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * 
 * Copyright 2019-2020, Mat Groves, All Rights Reserved
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var loaderBase = require('@pixi-spine/loader-base');
var base = require('@pixi-spine/base');
var loaders = require('@pixi/loaders');
var spine38 = require('@pixi-spine/runtime-3.8');
var spine37 = require('@pixi-spine/runtime-3.7');
var spine40 = require('@pixi-spine/runtime-4.0');

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () {
                        return e[k];
                    }
                });
            }
        });
    }
    n['default'] = e;
    return Object.freeze(n);
}

var spine38__namespace = /*#__PURE__*/_interopNamespace(spine38);
var spine37__namespace = /*#__PURE__*/_interopNamespace(spine37);
var spine40__namespace = /*#__PURE__*/_interopNamespace(spine40);

/**
 * @public
 */
 (function (SPINE_VERSION) {
    const UNKNOWN = 0; SPINE_VERSION[SPINE_VERSION["UNKNOWN"] = UNKNOWN] = "UNKNOWN";
    const VER37 = 37; SPINE_VERSION[SPINE_VERSION["VER37"] = VER37] = "VER37";
    const VER38 = 38; SPINE_VERSION[SPINE_VERSION["VER38"] = VER38] = "VER38";
    const VER40 = 40; SPINE_VERSION[SPINE_VERSION["VER40"] = VER40] = "VER40";
})(exports.SPINE_VERSION || (exports.SPINE_VERSION = {}));

/**
 * @public
 */
function detectSpineVersion(version) {
    const ver3 = version.substr(0, 3);
    const verNum = Math.floor(+ver3 * 10 + 1e-3);

    if (ver3 === '3.7') {
        return exports.SPINE_VERSION.VER37;
    }
    if (ver3 === '3.8') {
        return exports.SPINE_VERSION.VER38;
    }
    if (ver3 === '4.0') {
        return exports.SPINE_VERSION.VER40;
    }
    // try parse old versions with 3.7
    if (verNum < exports.SPINE_VERSION.VER37) {
        return exports.SPINE_VERSION.VER37;
    }
    return exports.SPINE_VERSION.UNKNOWN;
}

class UniBinaryParser  {constructor() { UniBinaryParser.prototype.__init.call(this); }
    __init() {this.scale = 1;}

    readSkeletonData(atlas, dataToParse) {
        let input = new base.BinaryInput(dataToParse);
        input.readString();
        let version = input.readString();
        let ver = detectSpineVersion(version);
        let parser = null;

        if (ver === exports.SPINE_VERSION.VER38) {
            parser = new spine38.SkeletonBinary(new spine38.AtlasAttachmentLoader(atlas));
        }

        input = new base.BinaryInput(dataToParse);
        input.readInt32();
        input.readInt32();
        version = input.readString();
        ver = detectSpineVersion(version);
        if (ver === exports.SPINE_VERSION.VER40) {
            parser = new spine40.SkeletonBinary(new spine40.AtlasAttachmentLoader(atlas));
        }
        if (!parser) {
            let error = `Unsupported version of spine model ${version}, please update pixi-spine`;
            console.error(error);
        }

        parser.scale = this.scale;
        return parser.readSkeletonData(dataToParse);
    }
}

class UniJsonParser  {constructor() { UniJsonParser.prototype.__init2.call(this); }
    __init2() {this.scale = 1;}

    readSkeletonData(atlas, dataToParse) {
        const version = dataToParse.skeleton.spine;
        const ver = detectSpineVersion(version);
        let parser = null;

        if (ver === exports.SPINE_VERSION.VER37) {
            parser = new spine37.SkeletonJson(new spine37.AtlasAttachmentLoader(atlas));
        }
        if (ver === exports.SPINE_VERSION.VER38) {
            parser = new spine38.SkeletonJson(new spine38.AtlasAttachmentLoader(atlas));
        }
        if (ver === exports.SPINE_VERSION.VER40) {
            parser = new spine40.SkeletonJson(new spine40.AtlasAttachmentLoader(atlas));
        }
        if (!parser) {
            let error = `Unsupported version of spine model ${version}, please update pixi-spine`;
            console.error(error);
        }

        parser.scale = this.scale;
        return parser.readSkeletonData(dataToParse);
    }
}

/**
 * @public
 */
class SpineParser extends loaderBase.AbstractSpineParser {
    createBinaryParser() {
        return new UniBinaryParser();
    }

    createJsonParser() {
        return new UniJsonParser();
    }

    parseData(resource, parser, atlas, dataToParse) {
        const parserCast = parser ;
        resource.spineData = parserCast.readSkeletonData(atlas, dataToParse);
        resource.spineAtlas = atlas;
    }

    static __initStatic() {this.use = new SpineParser().genMiddleware().use;}

    static registerLoaderPlugin() {
        loaders.Loader.registerPlugin(SpineParser);
    }
} SpineParser.__initStatic();

/**
 * @public
 */
class Spine extends base.SpineBase


 {

    createSkeleton(spineData) {
        const ver = detectSpineVersion(spineData.version);
        let spine = null;

        if (ver === exports.SPINE_VERSION.VER37) {
            spine = spine37__namespace;
        }
        if (ver === exports.SPINE_VERSION.VER38) {
            spine = spine38__namespace;
        }
        if (ver === exports.SPINE_VERSION.VER40) {
            spine = spine40__namespace;
        }
        if (!spine) {
            let error = `Cant detect version of spine model ${spineData.version}`;
            console.error(error);
        }
        this.skeleton = new spine.Skeleton(spineData);
        this.skeleton.updateWorldTransform();
        this.stateData = new spine.AnimationStateData(spineData);
        this.state = new spine.AnimationState(this.stateData);
    }
}

exports.Spine = Spine;
exports.SpineParser = SpineParser;
exports.detectSpineVersion = detectSpineVersion;
